File: src/types/mail.ts
export interface Mail {
  id: string;
  account_id: string;
  folder: string;
  message_id: string;
  in_reply_to: string | null;
  references: string | null;
  from_addr: string;
  to_addr: string;
  cc_addr: string | null;
  subject: string;
  body_text: string | null;
  body_html: string | null;
  date: string;
  has_attachments: boolean;
  raw_size: number | null;
  uid: number;
  flags: string | null;
  is_read: boolean;
  is_flagged: boolean;
  fetched_at: string;
  /** 以下2つは mails テーブルではなく mail_project_assignments 由来の注釈。
   * 未割り当てのメールでは null。確信度が中程度のAI分類に ⚠ を出すために使う */
  assigned_by?: string | null;
  confidence?: number | null;
}

export interface UnreadCounts {
  /** project_id → 未読件数 */
  by_project: Record<string, number>;
  /** 未分類メールの未読件数 */
  unclassified: number;
}

/** Tauri `send_mail` command の入力（Rust側 SendMailRequest と対応） */
export interface SendMailRequest {
  account_id: string;
  to: string[];
  cc: string[];
  bcc: string[];
  subject: string;
  body_text: string;
  reply_to_mail_id: string | null;
  /** リッチ本文のHTML。null ならプレーン送信（plain は Rust が HTML から生成） */
  body_html: string | null;
  /** 添付ファイルの絶対パス。Rust がバイト列を読み込む */
  attachments: string[];
}

/** ローカル下書き（v1: IMAP Draftsフォルダとの同期は将来） */
export interface Draft {
  id: string;
  account_id: string;
  to_addr: string;
  cc_addr: string;
  bcc_addr: string;
  subject: string;
  body_text: string;
  in_reply_to: string | null;
  created_at: string;
  updated_at: string;
}

/** Tauri `save_draft` command の入力（Rust側 SaveDraftRequest と対応） */
export interface SaveDraftRequest {
  id: string | null;
  account_id: string;
  to_addr: string;
  cc_addr: string;
  bcc_addr: string;
  subject: string;
  body_text: string;
  in_reply_to: string | null;
}

export interface SearchResult {
  mail: Mail;
  project_id: string | null;
  project_name: string | null;
  snippet: string;
}

export interface Thread {
  thread_id: string;
  subject: string;
  last_date: string;
  mail_count: number;
  from_addrs: string[];
  mails: Mail[];
  /** 集約表示（サブツリー展開）でのみ埋まる。選択ノード自身を除く所属案件の集合 */
  projects: ThreadProjectRef[];
}

/** 集約表示でスレッドに付ける「どの案件のメールか」の注釈 */
export interface ThreadProjectRef {
  project_id: string;
  /** 選択ノードからの相対パス（例: "埼玉 > 音響"） */
  display_path: string;
}

/** 一括操作（bulk_delete_mails 等）の結果。1件の失敗で残りは止めない */
export interface BulkResult {
  succeeded: string[];
  /** [mail_id, エラーメッセージ] の組 */
  failed: [string, string][];
}

/** backfill_account の結果 */
export interface BackfillOutcome {
  /** 今回取り込んだ件数 */
  fetched: number;
  /** これ以上サーバーに古いメールが無いか */
  exhausted: boolean;
}

File: src-tauri/src/models/mail.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Mail {
    pub id: String,
    pub account_id: String,
    pub folder: String,
    pub message_id: String,
    pub in_reply_to: Option<String>,
    pub references: Option<String>,
    pub from_addr: String,
    pub to_addr: String,
    pub cc_addr: Option<String>,
    pub subject: String,
    pub body_text: Option<String>,
    pub body_html: Option<String>,
    pub date: String,
    pub has_attachments: bool,
    pub raw_size: Option<i64>,
    pub uid: u32,
    pub flags: Option<String>,
    pub is_read: bool,
    pub is_flagged: bool,
    pub fetched_at: String,
    /// uid がサーバー実 UID として確定しているか。サーバー取得行は true、
    /// 送信時にローカル保存する Sent 行（推定 uid）は false。
    /// Sent 同期の watermark 計算（確定行のみの max uid）に使う。
    pub uid_confirmed: bool,
    /// 以下2つは `mails` テーブルのカラムではなく、`mail_project_assignments`
    /// から JOIN して載せる注釈。割り当てのないメールでは `None` になる。
    /// 確信度が中程度（`CONFIDENCE_UNCERTAIN`〜`CONFIDENCE_AUTO_ASSIGN`）の
    /// AI 分類にユーザー確認を促すため、UI まで運ぶ
    /// （設計: docs/design/2026-04-13-phase2-ai-classification-design.md）。
    #[serde(default)]
    pub assigned_by: Option<String>,
    #[serde(default)]
    pub confidence: Option<f64>,
}

/// アカウント内の未読件数の集計（folder='INBOX' のみ対象）
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnreadCounts {
    /// project_id → 未読件数
    pub by_project: std::collections::HashMap<String, u32>,
    /// 未分類メールの未読件数
    pub unclassified: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub mail: Mail,
    pub project_id: Option<String>,
    pub project_name: Option<String>,
    pub snippet: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Thread {
    pub thread_id: String,
    pub subject: String,
    pub last_date: String,
    pub mail_count: usize,
    pub from_addrs: Vec<String>,
    pub mails: Vec<Mail>,
    /// 集約表示（サブツリー展開）でのみ埋まる。メンバーメールの直接所属案件のうち、
    /// 選択ノード自身を除いたものの集合。選択ノード直属のみのスレッドは空。
    pub projects: Vec<ThreadProjectRef>,
}

/// スレッド一覧の1ページ分。
///
/// 一覧取得は必ず上限を持つ（ADR 0006 決定5）。切り出しの単位は「メール」ではなく
/// 「スレッド」——メール単位で LIMIT すると、同じスレッドの一部のメールだけが
/// 窓に入り、mail_count や参加者一覧が実データと食い違ったスレッドが UI に出る。
///
/// `has_more` は「この後ろにまだスレッドがあるか」。総件数は返さない
/// （COUNT(*) の全走査を毎回発生させないため）。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadPage {
    pub threads: Vec<Thread>,
    pub has_more: bool,
}

/// 集約表示でスレッドに付ける「どの案件のメールか」の注釈。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ThreadProjectRef {
    pub project_id: String,
    /// 選択ノードからの相対パス（例: 選択が「ツアー」なら "埼玉 > 音響"）。
    /// 階層内では同名案件が共存し得るため単一 name ではなくパスにする。
    pub display_path: String,
}
